cpp/ql/src/Critical/MissingCheckScanf.ql

```diff
@@ -20,81 +20,187 @@ import semmle.code.cpp.dataflow.DataFlow
-    if
-      exists(int value, int minGuard | minGuard = this.getMinimumGuardConstant() |
-        e.getBasicBlock() = blockGuardedBy(value, "==", call) and minGuard <= value
-        or
-        e.getBasicBlock() = blockGuardedBy(value, "<", call) and minGuard - 1 <= value
-        or
-        e.getBasicBlock() = blockGuardedBy(value, "<=", call) and minGuard <= value
-      )
-    then isGuarded = true
-    else isGuarded = false
-  }
-
-  /**
-   * Get a subsequent access of the same underlying storage,
-   * but before it gets reset or reused in another `scanf` call.
-   */
-  Access getAnAccess() {
-    exists(Instruction dst |
-      this.bigStep() = dst and
-      dst.getAst() = result and
-      valueNumber(dst) = valNum
+/**
+ * Holds if `call` is a `scanf`-like function that may write to `output` at index `index`.
+ *
+ * Furthermore, `instr` is the instruction that defines the address of the `index`'th argument
+ * of `call`, and `vn` is the value number of `instr.`
+ */
+predicate isSource(ScanfFunctionCall call, int index, Instruction instr, ValueNumber vn, Expr output) {
+  output = call.getOutputArgument(index).getFullyConverted() and
+  instr.getConvertedResultExpression() = output and
+  vn.getAnInstruction() = instr
+}
+
+/**
+ * Holds if `instr` is control-flow reachable in 0 or more steps from an a call
+ * to a call to a `scanf`-like function.
+ */
+pragma[nomagic]
+predicate fwdFlow0(Instruction instr) {
+  isSource(_, _, instr, _, _)
+  or
+  exists(Instruction prev |
+    fwdFlow0(prev) and
+    prev.getASuccessor() = instr
+  )
+}
+
+/**
+ * Holds if `instr` is part of the IR translation of `access` that
+ * is not an expression being deallocated, and `instr` has value
+ * number `vn`.
+ */
+predicate isSink(Instruction instr, Access access, ValueNumber vn) {
+  instr.getAst() = access and
+  not any(DeallocationExpr dealloc).getFreedExpr() = access and
+  vn.getAnInstruction() = instr
+}
+
+/**
+ * Holds if `instr` is part of a path from a call to a `scanf`-like function
+ * to a use of the written variable.
+ */
+pragma[nomagic]
+predicate revFlow0(Instruction instr) {
+  fwdFlow0(instr) and
+  (
+    isSink(instr, _, _)
+    or
+    exists(Instruction succ | revFlow0(succ) | instr.getASuccessor() = succ)
+  )
+}
+
+/**
+ * Holds if `instr` is part of a path from a call `to a `scanf`-like function
+ * that writes to a variable with value number `vn`, without passing through
+ * redefinitions of the variable.
+ */
+pragma[nomagic]
+private predicate fwdFlow(Instruction instr, ValueNumber vn) {
+  revFlow0(instr) and
+  (
+    isSource(_, _, instr, vn, _)
+    or
+    exists(Instruction prev |
+      fwdFlow(prev, vn) and
+      prev.getASuccessor() = instr and
+      not isBarrier(instr, vn)
+    )
+  )
+}
+
+/**
+ * Holds if `instr` is part of a path from a call `to a `scanf`-like function
+ * that writes to a variable with value number `vn`, without passing through
+ * redefinitions of the variable.
+ */
```

Comment:
revFlow and fwdFlow both have the same doc comment. I would point out the difference between them.

Reply:
Fixed in 59072f9e81d7878d34f0e3c45855cfeeebbe3524.